### PR TITLE
PYTHON-3776 add types to collection.py

### PR DESCRIPTION
**Pull request by sleepyStick**
https://jira.mongodb.org/browse/PYTHON-3776

### CODE REVIEW

File: pymongo/collection.py [hunk 1]
```diff
@@ -242,32 +262,37 @@ def __init__(
...
-    ):
+        sock_info: SocketInfo,
+        command: Mapping[str, Any],
+        read_preference: Optional[_ServerMode] = None,
+        codec_options: Optional[CodecOptions] = None,
+        check: bool = True,
+        allowable_errors: Optional[Container[Any]] = None,
+        read_concern: Optional[ReadConcern] = None,
+        write_concern: Optional[WriteConcern] = None,
+        collation: Union[Collation, Mapping[str, Any], None] = None,
```
Comment (blink1073) [inline pymongo/collection.py:285]: We have an alias for this: [`_CollationIn`](https://github.com/mongodb/mongo-python-driver/blob/94fabf5e988b02041693ff7f0921e4eee32ee9a0/pymongo/typings.py#L26) (for all usages in this file).

**Review by blink1073 — CHANGES_REQUESTED** [review]

**Review by blink1073 — APPROVED** [review]
LGTM overall, just one simplification

File: pymongo/helpers.py [hunk 2]
```diff
@@ -157,7 +158,7 @@ def _index_document(index_list: _IndexList) -> SON[str, Any]:
 def _check_command_response(
     response: Mapping[str, Any],
     max_wire_version: Optional[int],
-    allowable_errors: Optional[List[int]] = None,
+    allowable_errors: Optional[Container[int]] = None,
```
Comment (ShaneHarvey) [inline pymongo/helpers.py:161]: allowable_errors should be `Optional[Container[Union[int, str]]]`, for example:
```
pymongo/collection.py:2171:                allowable_errors=["ns not found", 26],
```

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: pymongo/collection.py [hunk 3]
```diff
@@ -104,10 +118,16 @@ class ReturnDocument:
 
 
 if TYPE_CHECKING:
+    from contextlib import _GeneratorContextManager
```
[the remark this answers is not in the document]
Comment (blink1073) [reply pymongo/collection.py:121]: Annoyingly, the return value of the `contextmanager` function in the [stdlib](https://github.com/python/typeshed/blob/31dca1d660937e2e3fe4670934b8c331d6f4144c/stdlib/contextlib.pyi#L76) has this return value.

[the remark this answers is not in the document]
Comment (blink1073) [reply pymongo/collection.py:121]: We're using `_GeneratorContextManager` in `server.py` already.

[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply pymongo/collection.py:121]: Nice find. I'd suggest using Generator here and in server.py since _GeneratorContextManager in private & undocumented.

**Review by ShaneHarvey — APPROVED** [review]

**Pull request merged**